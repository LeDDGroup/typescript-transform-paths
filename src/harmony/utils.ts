export type DownSampleTsTypes<TypeMap extends [unknown, unknown][], Tuple extends [...unknown[]]> = {
  [i in keyof Tuple]: Tuple[i] extends unknown[]
    ? DownSampleTsTypes<TypeMap, Tuple[i]>
    : DownSampleTsType<TypeMap, Tuple[i]>;
} & {
  length: Tuple["length"];
};

type DownSampleTsType<TypeMap extends [unknown, unknown][], T> =
  T extends Exclude<TypeMap[number][0], undefined> ? Extract<TypeMap[number], [T, unknown]>[1] : T;
