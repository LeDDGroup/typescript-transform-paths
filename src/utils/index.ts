export * from "./general-utils";
export * from "./elide-import-export";
export * from "./resolve-path-update-node";
export * from "./ts-type-conversion";
