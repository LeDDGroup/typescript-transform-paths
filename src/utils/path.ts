// noinspection JSUnusedLocalSymbols,JSUnusedAssignment

import path from "path";
import { normalizePath, Pattern, removePrefix, removeSuffix, ResolvedModuleFull } from "typescript";

/* ****************************************************************************************************************** */
// region: Types
/* ****************************************************************************************************************** */

export interface OutputPathDetail {
  isImplicitExtension: boolean
  isExternalLibraryImport: boolean
  resolvedExt: string | undefined;
  resolvedPath: string;
  outputPath: string | undefined;
  suppliedExt: string | undefined;
  implicitPath?: string;
  /** Package name from package.json */
  packageName?: string;
  /** Package name as written (could differ due to npm aliasing) */
  suppliedPackageName?: string
  suppliedPackagePath?: string
  packagePath?: string;
  tsPathMatch?: string;
}

// endregion

/* ****************************************************************************************************************** */
// region: Helpers
/* ****************************************************************************************************************** */

const pkgRegex = /^((@[^/]+\/[^/@]+)|([^@/]+))(?:\/([^@]+?))?$/;

/** @internal */
function getPaths(supplied: string, resolved: string) {
  let endMatchPos = 0;
  for (let i = 0; i < supplied.length && i < resolved.length; i++) {
    if (supplied[i] !== resolved[i]) {
      endMatchPos = i;
      break;
    }
  }

  const outputPath = supplied.slice(0, endMatchPos);
  const implicitPath = resolved.slice(endMatchPos);

  return { outputPath, implicitPath };
}

function getPackagePaths(moduleName: string, suppliedExt: string, subModule: string) {
  const { 1: suppliedPackageName, 2: suppliedPackagePath } = pkgRegex.exec(moduleName)!;

  const packagePathNoExt = fixupPartialPath(suppliedPackagePath, suppliedExt);

  if (!subModule) return { outputPath: packagePathNoExt };

  const subModuleNoExt = fixupPartialPath(subModule, path.extname(subModule));

  return {
    ...getPaths(packagePathNoExt, subModuleNoExt),
    suppliedPackageName,
    suppliedPackagePath
  };
}

function getModulePaths(
  moduleName: string,
  suppliedExt: string,
  resolvedFileName: string,
  resolvedExt: string,
  tsPathMatch: string | Pattern | undefined
) {
  // In this case, the file ending is a fixed path, so no further information needs to be determined
  if (typeof tsPathMatch === "string" || tsPathMatch?.suffix) return { outputPath: void 0 };

  const resolvedFileNameNoExt = fixupPartialPath(resolvedFileName, resolvedExt);
  const modulePathNoExt = tsPathMatch?.prefix
    ? fixupPartialPath(removePrefix(moduleName, tsPathMatch?.prefix), suppliedExt)
    : fixupPath(moduleName, suppliedExt);

  return getPaths(modulePathNoExt, resolvedFileNameNoExt);
}

function getTsPathMatch(match: string | Pattern): string {
  return typeof match === "string" ? match : [match.prefix, match.suffix].join("*");
}

// endregion

/* ****************************************************************************************************************** */
// region: Utils
/* ****************************************************************************************************************** */

export function joinPaths(...paths: (string | undefined)[]): string {
  return normalizePath(path.join('', ...paths.filter(p => typeof p === 'string') as string[]));
}

/**
 * Remove leading or trailing slashes
 * @p path to fix
 * @extName extname to remove
 */
export function fixupPartialPath(p: string, extName?: string): string {
  p = p.replace(/^[/\\]*(.+?)[/\\]*$/g, "$1");
  if (extName && p.slice(-extName.length) === extName) p = p.slice(0, p.length - extName.length);
  return p;
}

/**
 * Remove trailing slashes
 * @p path to fix
 * @extName extname to remove
 */
export function fixupPath(p: string, extName?: string): string {
  p = p.replace(/^(.+?)[/\\]*$/g, "$1");
  if (extName && p.slice(-extName.length) === extName) p = p.slice(0, p.length - extName.length);
  return p;
}

/** @internal â€” Uses internal TS type */
export function getOutputPathDetail(
  moduleName: string,
  resolvedModule: ResolvedModuleFull,
  pathMatch: string | Pattern | undefined
): OutputPathDetail {
  moduleName = fixupPath(moduleName);
  let suppliedExtName = path.extname(moduleName);
  const suppliedBaseName = path.basename(moduleName);
  let suppliedBaseNameNoExt = path.basename(moduleName, suppliedExtName);

  const resolvedFileName = resolvedModule.originalPath ?? resolvedModule.resolvedFileName;
  const resolvedExtName = path.extname(resolvedFileName);
  const resolvedFileNameNoExt = removeSuffix(resolvedFileName, resolvedExtName);
  const resolvedBaseName = path.basename(resolvedFileName);
  const resolvedBaseNameNoExt = path.basename(resolvedFileName, suppliedExtName);

  const tsPathMatch = pathMatch && getTsPathMatch(pathMatch);

  const { isExternalLibraryImport, packageId } = resolvedModule;
  const packageName = packageId?.name;
  const packageFileName = packageId && fixupPartialPath(packageId.subModuleName);
  const packageExtName = packageFileName && path.extname(packageFileName);
  const packageFileNameNoExt = packageFileName && removeSuffix(packageFileName, resolvedExtName);
  const packageBaseName = packageFileName && path.basename(packageFileName);
  const packageBaseNameNoExt = packageFileName && path.basename(packageFileName, packageExtName);

  const effectiveResolvedFileName = packageFileNameNoExt || resolvedFileNameNoExt;
  const effectiveResolvedBaseName = packageBaseName || resolvedBaseName;
  const effectiveResolvedBaseNameNoExt = packageBaseNameNoExt || resolvedBaseNameNoExt;
  const effectiveResolvedExtName = packageExtName || resolvedExtName;

  // Detect and fix invalid extname due to implicit ext (ie. `file.accounting.ts` could decide `accounting` is the extension)
  if (suppliedExtName && effectiveResolvedBaseNameNoExt && suppliedBaseName && effectiveResolvedBaseNameNoExt === suppliedBaseName) {
    suppliedBaseNameNoExt = suppliedBaseName;
    suppliedExtName = "";
  }

  const isImplicitExtension = !suppliedExtName;

  const pathDetail = resolvedModule.isExternalLibraryImport
    ? getPackagePaths(moduleName, suppliedExtName, packageId!.subModuleName)
    : getModulePaths(moduleName, suppliedExtName, resolvedFileName, resolvedExtName, tsPathMatch);

  return {
    isImplicitExtension,
    isExternalLibraryImport: !!isExternalLibraryImport,
    resolvedExt: effectiveResolvedExtName,
    suppliedExt: suppliedExtName,
    resolvedPath: effectiveResolvedFileName,
    tsPathMatch,
    packageName,
    packagePath: packageFileName,
    ...pathDetail,
  };
}

// endregion
