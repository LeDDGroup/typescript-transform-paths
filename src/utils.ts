import path from "path";
import url from 'url';

/* ****************************************************************************************************************** *
 * TS Exports
 * ****************************************************************************************************************** */

import { normalizePath } from 'typescript';
export { removeFileExtension, isAnySupportedFileExtension, removeSuffix } from 'typescript';
export { normalizePath };

/* ****************************************************************************************************************** */
// region: General Utils
/* ****************************************************************************************************************** */

export const isURL = (s: string): boolean => !!s && (!!url.parse(s).host || !!url.parse(s).hostname);
export const cast = <T>(v: any): T => v;
export const isBaseDir = (baseDir: string, testDir: string): boolean => {
  const relative = path.relative(baseDir, testDir);
  return relative ? !relative.startsWith("..") && !path.isAbsolute(relative) : true;
};
export const maybeAddRelativeLocalPrefix = (p: string) => (p[0] === "." ? p : `./${p}`);

// endregion

/* ****************************************************************************************************************** */
// region: Path Utils
/* ****************************************************************************************************************** */

/**
 * Merge and normalize paths (preserves relative prefix – ie. `./my/path`)
 */
export function joinPaths(...paths: (string | undefined)[]): string {
  // path.join / normalizePath cannot be used here, because they remove relative prefix
  return normalizeSlashes((paths.filter(p => !!p) as string[]).join('/'))!;
}

/**
 * Normalize slashes in path
 */
export function normalizeSlashes<T extends string | undefined>(p: T, opt?: { removeLeadingSlash?: boolean, removeTrailingSlash?: boolean }): T {
  if (!p) return p;

  let res = p.replace(/\\+|\/+/g, '/');

  if (opt?.removeLeadingSlash && res[0] === '/') res = res.slice(1);
  if (opt?.removeTrailingSlash && res[res.length-1] === '/') res = res.slice(0, res.length - 1);

  return res as T;
}

/**
 * Get extension for path (supports declaration style – ie. `d.ts`)
 */
export function extName(p: string) {
  const res = path.extname(p);
  const baseName = path.basename(p, res);
  return baseName.slice(-2) === '.d' ? `.d${res}` : res;
}

export function baseName(p: string, stripExtension: boolean | string = false) {
  const ext = !stripExtension ? void 0 : (typeof stripExtension === 'string' ? stripExtension : extName(p));
  return path.basename(p, ext);
}

export function resolvePath(...p: (string | undefined)[]): string {
  return normalizePath(path.resolve(joinPaths(...p))) as string;
}

export function dirName(p: string) {
  return normalizePath(path.dirname(p));
}

export function relativePath(from: string, to: string) {
  return normalizePath(path.relative(from, to));
}

// endregion
