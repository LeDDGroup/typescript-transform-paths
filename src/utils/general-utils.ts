import url from "url";
import path from "path";

/* ****************************************************************************************************************** *
 * General Utilities & Helpers
 * ****************************************************************************************************************** */

export const isURL = (s: string): boolean => !!s && (!!url.parse(s).host || !!url.parse(s).hostname);
export const cast = <T>(v: any): T => v;
export const isBaseDir = (baseDir: string, testDir: string): boolean => {
  const relative = path.relative(baseDir, testDir);
  return relative ? !relative.startsWith("..") && !path.isAbsolute(relative) : true;
};
export const maybeAddRelativeLocalPrefix = (p: string) => (p[0] === "." ? p : `./${p}`);
