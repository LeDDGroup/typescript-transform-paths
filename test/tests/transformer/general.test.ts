import path from "node:path";
import { createTsProgram, getEmitResultFromProgram, getRelativeEmittedFiles } from "../../utils";
import { ts, tsModules, projectsPaths } from "../../config";

describe(`Transformer -> General Tests`, () => {
  const projectRoot = path.join(projectsPaths, "general");
  const tsConfigFile = path.join(projectRoot, "tsconfig.json");

  test.each(tsModules)(`TypeScript %s`, (_description, tsInstance) => {
    const programWithTransformer = createTsProgram({ tsInstance: tsInstance as typeof ts, tsConfigFile });
    const transformedFiles = getEmitResultFromProgram(programWithTransformer);
    const emittedFiles = getRelativeEmittedFiles(projectRoot, transformedFiles);
    expect(emittedFiles).toMatchSnapshot();
  });
});
