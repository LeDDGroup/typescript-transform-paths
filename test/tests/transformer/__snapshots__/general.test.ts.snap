// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transformer -> General Tests TypeScript 3.6.5 1`] = `
{
  "circular/a.ts": {
    "dts": "import { B } from "./b";
export declare class A {
    name: string;
    constructor(name: string);
}
export declare const b: B;
",
    "js": "import { B } from "./b";
var A = /** @class */ (function () {
    function A(name) {
        this.name = name;
    }
    return A;
}());
export { A };
export var b = new B();
b.print(new A("This is a random name"));
",
  },
  "circular/b.ts": {
    "dts": "import { A } from "./a";
export declare class B {
    print(a: A): void;
}
",
    "js": "var B = /** @class */ (function () {
    function B() {
    }
    B.prototype.print = function (a) {
        console.log(a.name);
    };
    return B;
}());
export { B };
",
  },
  "core/index.ts": {
    "dts": "export { sum } from "../utils/index";
export { g } from "../secondary/hello";
export { sum as sum2 } from "../utils/sum";
export { NoRuntimecodeHere } from "../utils/types-only";
import "../circular/b";
",
    "js": "var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
export { sum } from "../utils/index";
export { g } from "../secondary/hello";
export { sum as sum2 } from "../utils/sum";
import { subs } from "../utils/index";
import "../circular/b";
import { A } from "../circular/a";
import * as path from "https://external.url/path.js";
import * as b from "../circular/a";
import * as c from "../circular/a";
import { myNative } from "../utils/utils.native";
c.A;
b.A;
path.sep;
myNative();
sum.sum(2, 3);
var n = null;
subs(2, 3);
var a = new A("");
(function () {
    return __awaiter(this, void 0, void 0, function () {
        var Logger, logger;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, import("../dynamic/logger")];
                case 1: return [4 /*yield*/, (_a.sent()).Logger];
                case 2:
                    Logger = _a.sent();
                    logger = new Logger();
                    logger.log("hi");
                    return [2 /*return*/];
            }
        });
    });
})();
(function () {
    return __awaiter(this, void 0, void 0, function () {
        var Tester, testerConst, testerClass;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, import("../dynamic/tester")];
                case 1:
                    Tester = (_a.sent()).Tester;
                    return [4 /*yield*/, import("../dynamic/tester")];
                case 2:
                    testerConst = (_a.sent()).tester;
                    testerClass = new Tester();
                    testerClass.test(12);
                    testerConst.test("12");
                    return [2 /*return*/];
            }
        });
    });
})();
",
  },
  "dynamic/logger.ts": {
    "dts": "declare type Tester = import("./tester").Tester;
export declare class Logger {
    level: string;
    tester: Tester;
    log(x: string): void;
}
export {};
",
    "js": "var Logger = /** @class */ (function () {
    function Logger() {
        this.level = "hi";
    }
    Logger.prototype.log = function (x) {
        console.log(x);
    };
    return Logger;
}());
export { Logger };
",
  },
  "dynamic/tester.ts": {
    "dts": "export declare class Tester {
    test(x: number): void;
}
export declare const tester: {
    test: (x: string) => void;
};
",
    "js": "var Tester = /** @class */ (function () {
    function Tester() {
    }
    Tester.prototype.test = function (x) {
        console.log(x);
    };
    return Tester;
}());
export { Tester };
export var tester = {
    test: function (x) { return console.log(x); },
};
",
  },
  "secondary/hello.ts": {
    "dts": "export declare const g = "hello";
",
    "js": "export var g = "hello";
",
  },
  "utils/index.ts": {
    "dts": "export * from "./sum";
export * from "./subs";
export { NoRuntimecodeHere } from "./types-only";
",
    "js": "export * from "./sum";
export * from "./subs";
",
  },
  "utils/subs.ts": {
    "dts": "export declare function subs(a: number, b: number): number;
",
    "js": "import { sum } from "./sum";
export function subs(a, b) {
    return sum(a, -b);
}
",
  },
  "utils/sum.ts": {
    "dts": "export declare function sum(a: number, b: number): number;
",
    "js": "export function sum(a, b) {
    return a + b;
}
",
  },
  "utils/types-only.ts": {
    "dts": "export declare type NoRuntimecodeHere = "never gonna give you up!";
",
    "js": "throw new Error("Not supposed to be!");
",
  },
  "utils/utils.native.ts": {
    "dts": "export declare function myNative(): void;
",
    "js": "export function myNative() { }
",
  },
}
`;

exports[`Transformer -> General Tests TypeScript 4.7.4 1`] = `
{
  "circular/a.ts": {
    "dts": "import { B } from "./b";
export declare class A {
    name: string;
    constructor(name: string);
}
export declare const b: B;
",
    "js": "import { B } from "./b";
var A = /** @class */ (function () {
    function A(name) {
        this.name = name;
    }
    return A;
}());
export { A };
export var b = new B();
b.print(new A("This is a random name"));
",
  },
  "circular/b.ts": {
    "dts": "import { A } from "./a";
export declare class B {
    print(a: A): void;
}
",
    "js": "var B = /** @class */ (function () {
    function B() {
    }
    B.prototype.print = function (a) {
        console.log(a.name);
    };
    return B;
}());
export { B };
",
  },
  "core/index.ts": {
    "dts": "export { sum } from "../utils/index";
export { g } from "../secondary/hello";
export { sum as sum2 } from "../utils/sum";
export { NoRuntimecodeHere } from "../utils/types-only";
import "../circular/b";
",
    "js": "var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
export { sum } from "../utils/index";
export { g } from "../secondary/hello";
export { sum as sum2 } from "../utils/sum";
import { subs } from "../utils/index";
import "../circular/b";
import { A } from "../circular/a";
import * as path from "https://external.url/path.js";
import * as b from "../circular/a";
import * as c from "../circular/a";
import { myNative } from "../utils/utils.native";
c.A;
b.A;
path.sep;
myNative();
sum.sum(2, 3);
var n = null;
subs(2, 3);
var a = new A("");
(function () {
    return __awaiter(this, void 0, void 0, function () {
        var Logger, logger;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, import("../dynamic/logger")];
                case 1: return [4 /*yield*/, (_a.sent()).Logger];
                case 2:
                    Logger = _a.sent();
                    logger = new Logger();
                    logger.log("hi");
                    return [2 /*return*/];
            }
        });
    });
})();
(function () {
    return __awaiter(this, void 0, void 0, function () {
        var Tester, testerConst, testerClass;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, import("../dynamic/tester")];
                case 1:
                    Tester = (_a.sent()).Tester;
                    return [4 /*yield*/, import("../dynamic/tester")];
                case 2:
                    testerConst = (_a.sent()).tester;
                    testerClass = new Tester();
                    testerClass.test(12);
                    testerConst.test("12");
                    return [2 /*return*/];
            }
        });
    });
})();
",
  },
  "dynamic/logger.ts": {
    "dts": "declare type Tester = import("./tester").Tester;
export declare class Logger {
    level: string;
    tester: Tester;
    log(x: string): void;
}
export {};
",
    "js": "var Logger = /** @class */ (function () {
    function Logger() {
        this.level = "hi";
    }
    Logger.prototype.log = function (x) {
        console.log(x);
    };
    return Logger;
}());
export { Logger };
",
  },
  "dynamic/tester.ts": {
    "dts": "export declare class Tester {
    test(x: number): void;
}
export declare const tester: {
    test: (x: string) => void;
};
",
    "js": "var Tester = /** @class */ (function () {
    function Tester() {
    }
    Tester.prototype.test = function (x) {
        console.log(x);
    };
    return Tester;
}());
export { Tester };
export var tester = {
    test: function (x) { return console.log(x); },
};
",
  },
  "secondary/hello.ts": {
    "dts": "export declare const g = "hello";
",
    "js": "export var g = "hello";
",
  },
  "utils/index.ts": {
    "dts": "export * from "./sum";
export * from "./subs";
export { NoRuntimecodeHere } from "./types-only";
",
    "js": "export * from "./sum";
export * from "./subs";
",
  },
  "utils/subs.ts": {
    "dts": "export declare function subs(a: number, b: number): number;
",
    "js": "import { sum } from "./sum";
export function subs(a, b) {
    return sum(a, -b);
}
",
  },
  "utils/sum.ts": {
    "dts": "export declare function sum(a: number, b: number): number;
",
    "js": "export function sum(a, b) {
    return a + b;
}
",
  },
  "utils/types-only.ts": {
    "dts": "export declare type NoRuntimecodeHere = "never gonna give you up!";
",
    "js": "throw new Error("Not supposed to be!");
export {};
",
  },
  "utils/utils.native.ts": {
    "dts": "export declare function myNative(): void;
",
    "js": "export function myNative() { }
",
  },
}
`;

exports[`Transformer -> General Tests TypeScript Latest 1`] = `
{
  "circular/a.ts": {
    "dts": "import { B } from "./b";
export declare class A {
    name: string;
    constructor(name: string);
}
export declare const b: B;
",
    "js": "import { B } from "./b";
var A = /** @class */ (function () {
    function A(name) {
        this.name = name;
    }
    return A;
}());
export { A };
export var b = new B();
b.print(new A("This is a random name"));
",
  },
  "circular/b.ts": {
    "dts": "import { A } from "./a";
export declare class B {
    print(a: A): void;
}
",
    "js": "var B = /** @class */ (function () {
    function B() {
    }
    B.prototype.print = function (a) {
        console.log(a.name);
    };
    return B;
}());
export { B };
",
  },
  "core/index.ts": {
    "dts": "export { sum } from "../utils/index";
export { g } from "../secondary/hello";
export { sum as sum2 } from "../utils/sum";
export { NoRuntimecodeHere } from "../utils/types-only";
import "../circular/b";
",
    "js": "var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
export { sum } from "../utils/index";
export { g } from "../secondary/hello";
export { sum as sum2 } from "../utils/sum";
import { subs } from "../utils/index";
import "../circular/b";
import { A } from "../circular/a";
import * as path from "https://external.url/path.js";
import * as b from "../circular/a";
import * as c from "../circular/a";
import { myNative } from "../utils/utils.native";
c.A;
b.A;
path.sep;
myNative();
sum.sum(2, 3);
var n = null;
subs(2, 3);
var a = new A("");
(function () {
    return __awaiter(this, void 0, void 0, function () {
        var Logger, logger;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, import("../dynamic/logger")];
                case 1: return [4 /*yield*/, (_a.sent()).Logger];
                case 2:
                    Logger = _a.sent();
                    logger = new Logger();
                    logger.log("hi");
                    return [2 /*return*/];
            }
        });
    });
})();
(function () {
    return __awaiter(this, void 0, void 0, function () {
        var Tester, testerConst, testerClass;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, import("../dynamic/tester")];
                case 1:
                    Tester = (_a.sent()).Tester;
                    return [4 /*yield*/, import("../dynamic/tester")];
                case 2:
                    testerConst = (_a.sent()).tester;
                    testerClass = new Tester();
                    testerClass.test(12);
                    testerConst.test("12");
                    return [2 /*return*/];
            }
        });
    });
})();
",
  },
  "dynamic/logger.ts": {
    "dts": "type Tester = import("./tester").Tester;
export declare class Logger {
    level: string;
    tester: Tester;
    log(x: string): void;
}
export {};
",
    "js": "var Logger = /** @class */ (function () {
    function Logger() {
        this.level = "hi";
    }
    Logger.prototype.log = function (x) {
        console.log(x);
    };
    return Logger;
}());
export { Logger };
",
  },
  "dynamic/tester.ts": {
    "dts": "export declare class Tester {
    test(x: number): void;
}
export declare const tester: {
    test: (x: string) => void;
};
",
    "js": "var Tester = /** @class */ (function () {
    function Tester() {
    }
    Tester.prototype.test = function (x) {
        console.log(x);
    };
    return Tester;
}());
export { Tester };
export var tester = {
    test: function (x) { return console.log(x); },
};
",
  },
  "secondary/hello.ts": {
    "dts": "export declare const g = "hello";
",
    "js": "export var g = "hello";
",
  },
  "utils/index.ts": {
    "dts": "export * from "./sum";
export * from "./subs";
export { NoRuntimecodeHere } from "./types-only";
",
    "js": "export * from "./sum";
export * from "./subs";
",
  },
  "utils/subs.ts": {
    "dts": "export declare function subs(a: number, b: number): number;
",
    "js": "import { sum } from "./sum";
export function subs(a, b) {
    return sum(a, -b);
}
",
  },
  "utils/sum.ts": {
    "dts": "export declare function sum(a: number, b: number): number;
",
    "js": "export function sum(a, b) {
    return a + b;
}
",
  },
  "utils/types-only.ts": {
    "dts": "export type NoRuntimecodeHere = "never gonna give you up!";
",
    "js": "throw new Error("Not supposed to be!");
export {};
",
  },
  "utils/utils.native.ts": {
    "dts": "export declare function myNative(): void;
",
    "js": "export function myNative() { }
",
  },
}
`;
